File: rust/src/problems/p148.rs
use crate::utils::linked_list::ListNode;

pub struct Solution;

impl Solution {
    pub fn sort_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        fn split(
            mut head: Option<Box<ListNode>>,
            n: usize,
        ) -> (Option<Box<ListNode>>, Option<Box<ListNode>>) {
            let mut h = &mut head;
            let mut i = 0;
            while i < (n + 1) / 2 {
                h = &mut h.as_mut().unwrap().next;
                i += 1;
            }
            let h2 = h.take();
            (head, h2)
        }

        fn merge(
            mut h1: Option<Box<ListNode>>,
            mut h2: Option<Box<ListNode>>,
        ) -> Option<Box<ListNode>> {
            let mut h = ListNode::new(0);
            let mut tail = &mut h.next;
            while h1.is_some() && h2.is_some() {
                let v1 = h1.as_ref().unwrap().val;
                let v2 = h2.as_ref().unwrap().val;
                if v1 < v2 {
                    let h1_next = h1.as_mut().unwrap().next.take();
                    *tail = h1;
                    h1 = h1_next;
                } else {
                    let h2_next = h2.as_mut().unwrap().next.take();
                    *tail = h2;
                    h2 = h2_next;
                }
                tail = &mut tail.as_mut().unwrap().next;
            }
            if h1.is_some() {
                *tail = h1;
            } else if h2.is_some() {
                *tail = h2;
            }
            h.next
        }

        fn sort(head: Option<Box<ListNode>>, n: usize) -> Option<Box<ListNode>> {
            if n <= 1 {
                head
            } else {
                let (h1, h2) = split(head, n);
                let h1 = sort(h1, (n + 1) / 2);
                let h2 = sort(h2, n / 2);
                merge(h1, h2)
            }
        }

        let mut n = 0;
        let mut h = &head;
        while let Some(node) = h {
            n += 1;
            h = &node.next;
        }
        sort(head, n)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn case1() {
        let head = ListNode::from_vec(&[4, 2, 1, 3]);
        assert_eq!(ListNode::from_vec(&[1, 2, 3, 4]), Solution::sort_list(head));
    }
}

File: cpp/p148.cpp
#include <algorithm>
#include <cassert>
#include <vector>

#include "helpers.h"

using namespace std;

class Solution {
    ListNode *split(ListNode *head, int size) {
        for (int i = 1; i < size && head; i++) {
            head = head->next;
        }
        if (!head || !head->next) {
            return nullptr;
        } else {
            auto tail = head->next;
            head->next = nullptr;
            return tail;
        }
    }
    pair<ListNode *, ListNode *> merge(ListNode *l1, ListNode *l2) {
        ListNode d(0), *h = &d;
        while (l1 && l2) {
            if (l1->val <= l2->val) {
                h->next = l1;
                l1 = l1->next;
            } else {
                h->next = l2;
                l2 = l2->next;
            }
            h = h->next;
        }
        h->next = l1 ? l1 : l2;
        while (h->next) {
            h = h->next;
        }
        return {d.next, h};
    }

public:
    ListNode *sortList(ListNode *head) {
        int n = 0;
        ListNode *h = head;
        while (h) {
            n += 1;
            h = h->next;
        }
        ListNode d(0, head);
        for (int i = 1; i < n; i *= 2) {
            ListNode *prev = &d;
            h = d.next;
            while (h) {
                ListNode *l1 = h;
                ListNode *l2 = split(l1, i);
                h = split(l2, i);
                auto [head, tail] = merge(l1, l2);
                prev->next = head;
                prev = tail;
            }
        }
        return d.next;
    }
};

int main() {
    auto r = Solution().sortList(parse_singly_linked_list({4, 2, 1, 3}));
    ListNode *expected = parse_singly_linked_list({1, 2, 3, 4});
    while (expected) {
        assert(r && r->val == expected->val);
        expected = expected->next;
        r = r->next;
    }
    assert(!r);
    return 0;
}